File: Sources/AINotebookCore/AttachmentStore.swift
import Foundation
import GRDB

@MainActor
public final class AttachmentStore {

    private let store: NotebookStore
    public let root: URL

    public init(store: NotebookStore, root: URL) {
        self.store = store
        self.root = root
        try? FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
    }

    public static func defaultRoot(fileManager: FileManager = .default) throws -> URL {
        let appSupport = try fileManager.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask, appropriateFor: nil, create: true
        )
        let container = appSupport.appendingPathComponent("AINotebook", isDirectory: true)
        let attachments = container.appendingPathComponent("attachments", isDirectory: true)
        try fileManager.createDirectory(at: attachments, withIntermediateDirectories: true)
        return attachments
    }

    @discardableResult
    public func save(
        noteId: Int64,
        noteUuid: String,
        filename: String,
        mime: String,
        bytes: Data
    ) throws -> NoteAttachment {
        let folder = root.appendingPathComponent(noteUuid, isDirectory: true)
        try FileManager.default.createDirectory(at: folder, withIntermediateDirectories: true)
        let resolved = uniqueFilename(in: folder, requested: filename)
        let url = folder.appendingPathComponent(resolved)
        try bytes.write(to: url)
        var att = NoteAttachment(
            noteId: noteId,
            noteUuid: noteUuid,
            filename: resolved,
            mime: mime,
            byteSize: Int64(bytes.count)
        )
        try store.runOnDatabase { db in
            try att.insert(db)
        }
        return att
    }

    public func read(noteUuid: String, filename: String) throws -> Data {
        let url = root
            .appendingPathComponent(noteUuid, isDirectory: true)
            .appendingPathComponent(filename)
        return try Data(contentsOf: url)
    }

    public func list(noteId: Int64) throws -> [NoteAttachment] {
        try store.runOnDatabase { db in
            try NoteAttachment
                .filter(NoteAttachment.Columns.noteId.column == noteId)
                .order(NoteAttachment.Columns.createdAt.column.asc)
                .fetchAll(db)
        }
    }

    public func deleteFolder(noteUuid: String) throws {
        let folder = root.appendingPathComponent(noteUuid, isDirectory: true)
        if FileManager.default.fileExists(atPath: folder.path) {
            try FileManager.default.removeItem(at: folder)
        }
    }

    private func uniqueFilename(in folder: URL, requested: String) -> String {
        let stem = (requested as NSString).deletingPathExtension
        let ext = (requested as NSString).pathExtension
        let dotExt = ext.isEmpty ? "" : ".\(ext)"
        var candidate = requested
        var n = 2
        while FileManager.default.fileExists(atPath: folder.appendingPathComponent(candidate).path) {
            candidate = "\(stem) (\(n))\(dotExt)"
            n += 1
            if n > 9_999 { break }
        }
        return candidate
    }
}

File: windows/src/AINotebook.Core/Storage/AttachmentStore.cs
using AINotebook.Core.Models;
using Dapper;

namespace AINotebook.Core.Storage;

/// <summary>
/// Stores attachment bytes on disk under root/&lt;noteUuid&gt;/&lt;filename&gt;
/// and a row in the attachments table. Faithful to AttachmentStore.swift.
/// </summary>
public sealed class AttachmentStore
{
    private readonly NotebookStore _store;
    public string Root { get; }

    public AttachmentStore(NotebookStore store, string root)
    {
        _store = store;
        Root = root;
        Directory.CreateDirectory(root);
    }

    /// <summary>%APPDATA%\AINotebook\attachments, created on demand.</summary>
    public static string DefaultRoot()
    {
        var appData = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
        var dir = Path.Combine(appData, "AINotebook", "attachments");
        Directory.CreateDirectory(dir);
        return dir;
    }

    public NoteAttachment Save(long noteId, string noteUuid, string filename, string mime, byte[] bytes)
    {
        var folder = Path.Combine(Root, noteUuid);
        Directory.CreateDirectory(folder);
        var resolved = UniqueFilename(folder, filename);
        File.WriteAllBytes(Path.Combine(folder, resolved), bytes);

        var now = DateTime.UtcNow;
        var id = _store.Connection.ExecuteScalar<long>(
            """
            INSERT INTO attachments(note_id, note_uuid, filename, mime, byte_size, created_at)
            VALUES($nid, $uuid, $file, $mime, $size, $created);
            SELECT last_insert_rowid();
            """,
            new { nid = noteId, uuid = noteUuid, file = resolved, mime, size = (long)bytes.Length, created = SqliteDate.ToDb(now) });
        return new NoteAttachment(id, noteId, noteUuid, resolved, mime, bytes.Length, now);
    }

    public byte[] Read(string noteUuid, string filename) =>
        File.ReadAllBytes(Path.Combine(Root, noteUuid, filename));

    public IReadOnlyList<NoteAttachment> List(long noteId) =>
        _store.Connection.Query(
            "SELECT id, note_id, note_uuid, filename, mime, byte_size, created_at FROM attachments WHERE note_id=$nid ORDER BY created_at ASC",
            new { nid = noteId })
            .Select(r => new NoteAttachment(
                (long)r.id, (long)r.note_id, (string)r.note_uuid, (string)r.filename,
                (string)r.mime, (long)r.byte_size, SqliteDate.FromDb((string)r.created_at)))
            .ToList();

    public void DeleteFolder(string noteUuid)
    {
        var folder = Path.Combine(Root, noteUuid);
        if (Directory.Exists(folder)) Directory.Delete(folder, recursive: true);
    }

    private static string UniqueFilename(string folder, string requested)
    {
        var stem = Path.GetFileNameWithoutExtension(requested);
        var ext = Path.GetExtension(requested); // includes leading '.' or "" if none
        var candidate = requested;
        var n = 2;
        while (File.Exists(Path.Combine(folder, candidate)))
        {
            candidate = $"{stem} ({n}){ext}";
            n++;
            if (n > 9999) break;
        }
        return candidate;
    }
}
